### ISSUE TITLE
3.0 Messages of notEmpty and allowEmpty validation rules not set in Validator

### ISSUE THREAD

**Issue by fub-frank** [cakephp/cakephp#5856]
If the `notEmpty` or `allowEmpty` validation rules are added by using the `add` method of the Validator instead of the `notEmpty` or `allowEmpty` methods the custom error message is not set and gets lost.

```php
// in a Table Model:
public function validationDefault(Validator $validator)
{
    $validator = parent::validationDefault($validator);
    $validator->add('some_field_name', [
            'notEmpty' => [
                'rule' => 'notEmpty',
                'message' => 'a custom error message'
            ]
        ]);

    return $validator;
}
```

The custom error message is lost.
If the `notEmpty` or `allowEmpty` methods are used, the custom error message will be saved into the array `$_allowEmptyMessages[$field]` of the `Cake\Validation\Validator` class.
```php
// Cake\Validation\Validator
public function notEmpty($field, $message = null, $when = false)
{
    if ($when === 'create' || $when === 'update') {
        $when = $when === 'create' ? 'update' : 'create';
    } elseif (is_callable($when)) {
        $when = function ($context) use ($when) {
            return !$when($context);
        };
    }

    $this->field($field)->isEmptyAllowed($when);
    if ($message) {
        $this->_allowEmptyMessages[$field] = $message;
    }
    return $this;
}
```

But this is not the case, if using the `add` method
```php
// Cake\Validation\Validator
public function add($field, $name, $rule = [])
{
    $field = $this->field($field);

    if (!is_array($name)) {
        $rules = [$name => $rule];
    } else {
        $rules = $name;
    }

    foreach ($rules as $name => $rule) {
        $field->add($name, $rule);
    }

    return $this;
}
```

When the errors are returned the custom error message is not set in `$_allowEmptyMessages[$field]` and the default error message `This field cannot be left empty` will be returned
```php
// Cake\Validation\Validator

public function errors(array $data, $newRecord = true)
{
    // ...
    $emptyMessage = 'This field cannot be left empty';
    // ...
    if (!$canBeEmpty && $isEmpty) {
        $errors[$name][] = isset($this->_allowEmptyMessages[$name])
            ? $this->_allowEmptyMessages[$name]
            : $emptyMessage;
        continue;
    }
}
```

Here is a quick fix for the `add` Method, that adds the message to the `$_allowEmptyMessages[$field]` array. But maybe there is a better solution to this or i missed something:
```php
// Cake\Validation\Validator
public function add($field, $name, $rule = [])
{
    $fieldName = $field;
    $field = $this->field($field);

    if (!is_array($name)) {
        $rules = [$name => $rule];
    } else {
        $rules = $name;
    }

    foreach ($rules as $name => $rule) {
        if (($name == 'notEmpty' || $name == 'allowEmpty') && isset($rule['message'])) {
            $this->_allowEmptyMessages[$fieldName] = $rule['message'];
        }
        $field->add($name, $rule);
    }

    return $this;
}
```

**Comment by jadb** [comment]
Do you have a failing test for that?

**Comment by fub-frank** [comment]
Here is a quick failing test case for that. The test case is for `notEmpty` but the same applys to `allowEmpty`

```php
<?php
namespace App\Test\TestCase\Model;

use Cake\TestSuite\TestCase;
use Cake\Validation\Validator;

class ProgressHelperTest extends TestCase
{

    /**
     * This test case will work as expected because the
     * notEmpty validation is added to the validator by using
     * $validator->notEmpty().
     */
    public function testNotEmptyWithNotEmpty()
    {
        $expectedErrorMessage = 'non standard error message';


        $validator = new Validator();
        $validator->notEmpty('password', $expectedErrorMessage);

        $data = ['password' => ''];
        $errors = $validator->errors($data, true);
        $actualErrorMessage = $errors['password'][0];

        $this->assertTextEquals($expectedErrorMessage, $actualErrorMessage);
    }

    /**
     * This test case will fail. The notEmpty rule is added to the
     * $validator using the $validator->add() method. The custom
     * error message will be lost and the default error message
     * will be returned by the $validator->errors() method.
     */
    public function testNotEmptyWithAdd()
    {
        $expectedErrorMessage = 'non standard error message';


        $validator = new Validator();
        $validator
            ->add('password', [
                'notEmpty' => [
                    'rule' => 'notEmpty',
                    'message' => $expectedErrorMessage
                ]
            ])
        ;

        $data = ['password' => ''];
        $errors = $validator->errors($data, true);
        $actualErrorMessage = $errors['password'][0];

        // this will fail
        $this->assertTextEquals($expectedErrorMessage, $actualErrorMessage);
    }
}
```

**Comment by markstory** [comment]
Why do you want to use add() instead of notEmpty()?

**Comment by fub-frank** [comment]
because it is one of many rules that i want to add to one field. It is faster to just write a single array with all rules and pass it to the `add` method instead of using the `add` method for some of the rules and then the `notEmpty` method on the same field. Also the manual states that one can add many rules at once using the `add` method and does not state that `notEmpty` / `allowEmpty` are exceptions to this.

If you don't want to fix this at least exclude `notEmpty` and `alloEmpty` from the valid arguments to the `add` method and throw an exception. Yet it would be odd to allow some of the validation rules to be added by `add` and some not.

**Comment by markstory** [comment]
@fub-frank 'allowEmpty' is not valid as rule for `add()`.  The 'notEmpty' validator only exists for backwards compatibility and still needs to be combined with `notEmpty()` to actually work.

**Comment by robertpustulka** [comment]
Isn't it related to #5785 ?

**Comment by ADmad** [comment]
@robertpustulka Yup, now that you mention it is same. We failed to notice it earlier.

TLDR:  Don't add `notEmpty` as rule since it cause `Validator::notEmpty()` to be used instead of `Validation::notEmpty()` as one would expect.

**Comment by markstory** [comment]
Closing as Validation::notEmpty() should generally not be used. I've updated the API docs in 2091c688316bf1637c6ab5f7e69af5316d9dd6ef.
